### DISCUSSION TITLE
Better Typescript Examples?

### DISCUSSION THREAD

**Discussion by mikecann** [reimagined/resolve#1750] [General]
I asked this question on Stack Overfow too (https://stackoverflow.com/questions/66350824/are-there-any-good-modern-typescript-examples-for-resolvejs) but I think it would be better to have it as more of a disussion.

Roman Eremin mentioned that there are better Typescript definitions coming soon but arent prioritising Typescript examples right now. IMO this is a big miss because Typescript is really a big force in large complex projects these days and I can imagine that a lot of Resolve users (like myself) will be Typescript first and wont even approach a project unless it is Typescript based. 

IMO some of the more complex examples such as the Hacker News example should be converted to Typescript. Im willing to give it a go once the new type defs have landed if you agree..

**Comment by superroma** [comment]
Hi Mike,

Welcome to reSolve discussions, nice to meet you here!

See, I don't want us end up with two sets of examples to maintain. Our hypothesis was that any TS developer can understand JS and write similar code in TS, so we can live with JS examples and have only one TS template for new empty TS project.

On the other hand, I guess you are right and not mentioning TS anywhere, can turn some TS developers away from reSolve. Internally reSolve was started as JS project (TS at that poing was all-or-nothing), but is being converted to TS recently. So TS examples are natural for the team.

There is an obvious option to write example in TS and then just use TS compiler to create a pure JS example. We need to investigate if this process can be automated and produce a clean readable JS code and project.  

Let us research this after release, I hope we can do this early next week.

Also, don't forget that we have a documentation with code samples. Is it a problem for you as TS developer to see pure JS code samples? (For me - not a problem, but i'm not an end-user)

**Reply by mikecann** [reply]
> Let us research this after release, I hope we can do this early next week.

Fantastic, thankyou for being so open to community suggestions like this. Im really excited by resolve having built a purely serverless EventSourcing solution a few years back (when there was virtually no info on the topic) and so can appreciate all the hard work you have put in here.

> Is it a problem for you as TS developer to see pure JS code samples? (For me - not a problem, but i'm not an end-user)

Not massively but I do personally give much more weight to a project that can show that they have thought hard about the types and good experience for TS devs. 

One thing I love to see is multi-select boxes that let you see it in JS or TS (or other languages). Probably there are tools out there that automatically generate the example JS from the TS examples..

![image](https://user-images.githubusercontent.com/215033/110777952-afbf7000-829c-11eb-972d-a38334efe741.png)

**Reply by mikecann** [reply]
@superroma hey dont mean to pester but has there been any movement on the Typescript front? Keen to explore Resolve more but with a lack of typescript examples its hard to know what the intended way to write a resolve app using typescript is.

**Reply by superroma** [reply]
No TS examples yet.
Good news is that we have released v 0.30, so now we have now some time to spend on housekeeping. 
TS examples are on the list, we are currently testing the pipeline to convert TS to JS (so we don't have to maintain two sets of examples).

**Reply by mikecann** [reply]
Fanstastic! I look forward to it :)

**Reply by max-vasin** [reply]
Hi, @mikecann. We've released [new](https://github.com/reimagined/resolve/releases/tag/V0.31.0) version of the framework with reworked examples. Every example now originally written in TypeScript and automatically converted to JavaScript. You can use **create-resolve-app tool** with *--typescript* flag or browse manually in repository.

**Reply by mikecann** [reply]
@max-vasin I saw that, incredibe work guys, I look forward to having a tinker with it soon!
